revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,259 @@
+		reader := bytes.NewReader(ocspRequest)
+		resp, err = opts.HTTPClient.Post(server, "application/ocsp-request", reader)
+	} else {
+		encodedReq := base64.URLEncoding.EncodeToString(ocspRequest)
+		var reqURL string
+		reqURL, err = url.JoinPath(server, encodedReq)
+		if err != nil {
+			return nil, OCSPCheckError{Err: err}
+		}
+		resp, err = opts.HTTPClient.Get(reqURL)
+	}
+
+	if err != nil {
+		var urlErr *url.Error
+		if errors.As(err, &urlErr) && urlErr.Timeout() {
+			return nil, TimeoutError{}
+		}
+		return nil, OCSPCheckError{Err: err}
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, fmt.Errorf("failed to retrieve OSCP: response had status code %d", resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(io.LimitReader(resp.Body, ocspMaxResponseSize))
+	if err != nil {
+		return nil, OCSPCheckError{Err: err}
+	}
+
+	switch {
+	case bytes.Equal(body, ocsp.UnauthorizedErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP unauthorized")}
+	case bytes.Equal(body, ocsp.MalformedRequestErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP malformed")}
+	case bytes.Equal(body, ocsp.InternalErrorErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP internal error")}
+	case bytes.Equal(body, ocsp.TryLaterErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP try later")}
+	case bytes.Equal(body, ocsp.SigRequredErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP signature required")}
+	}
+
+	return ocsp.ParseResponseForCert(body, cert, issuer)
+}
+
+func errToServerResult(err error) *base.ServerResult {
+	if err == nil {
+		return &base.ServerResult{
+			Result: base.ResultOK,
+			Error:  nil,
+		}
+	} else if errors.Is(err, NoOCSPServerError{}) || errors.Is(err, PKIXNoCheckError{}) {
+		return &base.ServerResult{
+			Result: base.ResultNonRevokable,
+			Error:  nil,
+		}
+	} else if errors.Is(err, RevokedError{}) {
+		return &base.ServerResult{
+			Result: base.ResultRevoked,
+			Error:  err,
+		}
+	}
+	// Includes OCSPCheckError, UnknownStatusError, base.InvalidChainError, and
+	// TimeoutError
+	return &base.ServerResult{
+		Result: base.ResultUnknown,
+		Error:  err,
+	}
+}
+
+func serverResultsToCertRevocationResult(serverResults []*base.ServerResult) *base.CertRevocationResult {
+	if len(serverResults) == 1 {
+		return &base.CertRevocationResult{
+			Result:        serverResults[0].Result,
+			ServerResults: []*base.ServerResult{serverResults[0]},
+		}
+	}
+	currResult := base.ResultOK
+	for _, serverResult := range serverResults {
+		switch serverResult.Result {
+		case base.ResultRevoked:
+			currResult = base.ResultRevoked
+		case base.ResultUnknown:
+			if currResult != base.ResultRevoked {
+				currResult = base.ResultUnknown
+			}
+		case base.ResultNonRevokable:
+			if currResult != base.ResultRevoked && currResult != base.ResultUnknown {
+				currResult = base.ResultNonRevokable
+			}
+		}
+	}
```

Comment:
3. From our previous discussion, we decided for the PKIXNoCheckError to be associated with a different status than unknown or OK, and decided to use NonRevokable for it (since it was already being used for NoOCSPServerErrors and root certs). However, PKIXNoCheckError is specific to individual server responses. If we make a new result type just for PKIXNoCheckError or certs with CRL but not OCSP (maybe something like CRLRequired), then this can be changed to that new result.

1 & 2: There are two potential results for ServerErrors whenever there is more than one (an Unknown result, and a NonRevokable result for PKIXNoCheckErrors). If all of the servers have PkixNoCheckErrors, then the result for that cert should be NonRevokable (as the responder cannot be verified without a CRL check). However, if there is any other error, it will be Unknown. Thus, since there are two potential results, it doesn't make sense to use the last one as it may not be representative of all the others that came before it. Additionally, this is used in other scenarios (like tests) to easily convert any array of ServerResults into a CertResult, so it is intended to work for any set of results, regardless of what we'd anticipate for the ServerResults in certCheckStatus.

Reply:
This has been updated to use last result after removing PKIXNoCheckError so that all results should be Unknown (meaning there would be no potential loss of information by using last result)